From fcd5d94fd601b97c8934c5fe90cd20d67877d79f Mon Sep 17 00:00:00 2001
From: "zhe.wang" <zhe.wang@amlogic.com>
Date: Thu, 15 Jan 2015 11:32:37 +0800
Subject: [PATCH 5329/5965] PD#102274: when amaudio2 is opened, set i2s and
 iec958 buffer the same source

Squashed commit of the following:

commit 918631228ebae7bf21bbfb422ebf660b20cb7447
Author: zhe.wang <zhe.wang@amlogic.com>
Date:   Wed Jan 14 15:34:58 2015 +0800

    PD#102274: M6 platform has compile error, move the function to amaudio2 driver file

    Change-Id: I01da2083f455c610ed03b42ba8bb5ca3b02bdffd

commit 08690a9099a8ea342c18584051843f4647174d4c
Author: zhe.wang <zhe.wang@amlogic.com>
Date:   Wed Jan 14 14:29:52 2015 +0800

    PD#102274: when amaudio2 is opened, set i2s and iec958 buffer the same source

    Change-Id: Ic3e655b6adff0c1e9e50c3bc90526349cf7d4cdc

Change-Id: Icdde55971b01c23b4a875b0e7d7641f2f19c8488
---
 drivers/amlogic/amaudio2/amaudio2.c | 12 +++++++++++-
 sound/soc/aml/m8/aml_i2s_dai.c      |  8 ++++++--
 2 files changed, 17 insertions(+), 3 deletions(-)

diff --git a/drivers/amlogic/amaudio2/amaudio2.c b/drivers/amlogic/amaudio2/amaudio2.c
index f84f4f0a1baf..76a2488eed54 100755
--- a/drivers/amlogic/amaudio2/amaudio2.c
+++ b/drivers/amlogic/amaudio2/amaudio2.c
@@ -105,7 +105,6 @@ static int music_gain = 128;
 static int audio_out_mode = 0;
 static int audio_out_read_enable = 0;
 int amaudio2_enable = 0;
-
 static irqreturn_t i2s_out_callback(int irq, void* data);
 static unsigned get_i2s_out_size(void);
 static unsigned get_i2s_out_ptr(void);
@@ -717,6 +716,15 @@ static ssize_t store_audio_read_enable(struct class* class, struct class_attribu
   	return count;
 }
 
+int set_i2s_iec958_samesource(int enable){
+    if(enable == 0){
+        WRITE_MPEG_REG_BITS(AIU_I2S_MISC, 0, 3, 1);
+    }else if (enable == 1){
+        WRITE_MPEG_REG_BITS(AIU_I2S_MISC, 1, 3, 1);
+    }
+    return 0;
+}
+
 static ssize_t show_aml_amaudio2_enable(struct class* class, struct class_attribute* attr,
     char* buf)
 {
@@ -729,9 +737,11 @@ static ssize_t store_aml_amaudio2_enable(struct class* class, struct class_attri
 	if(buf[0] == '0'){
 		printk(KERN_INFO "amaudio2 is disable!\n");
 		amaudio2_enable = 0;
+		set_i2s_iec958_samesource(0);
 	}else if(buf[0] == '1'){
 		printk(KERN_INFO "amaudio2 is enable!\n");
 		amaudio2_enable = 1;
+		set_i2s_iec958_samesource(1);
 	}else{
 		printk(KERN_INFO "Invalid argument!\n");
 	}
diff --git a/sound/soc/aml/m8/aml_i2s_dai.c b/sound/soc/aml/m8/aml_i2s_dai.c
index 57955e7224fb..614da2792c38 100755
--- a/sound/soc/aml/m8/aml_i2s_dai.c
+++ b/sound/soc/aml/m8/aml_i2s_dai.c
@@ -37,12 +37,13 @@ static int i2s_pos_sync = 0;
 #define ALSA_DEBUG(fmt,args...) 
 #define ALSA_TRACE()   
 #endif
-
 extern int amaudio2_enable;
+extern int set_i2s_iec958_samesource(int enable);
+
 static int i2sbuf[32+16];
 static void  aml_i2s_play(void)
 {
-        audio_util_set_dac_i2s_format(AUDIO_ALGOUT_DAC_FORMAT_DSP); 	
+	audio_util_set_dac_i2s_format(AUDIO_ALGOUT_DAC_FORMAT_DSP);
 	audio_set_i2s_mode(AIU_I2S_MODE_PCM16);
 	memset(i2sbuf,0,sizeof(i2sbuf));	
 	audio_set_aiubuf((virt_to_phys(i2sbuf)+63)&(~63),128,2);
@@ -193,6 +194,7 @@ static int aml_dai_i2s_prepare(struct snd_pcm_substream *substream,
         aml_hw_i2s_init(runtime);
         if(amaudio2_enable == 1){
             aml_hw_iec958_init(substream);
+            set_i2s_iec958_samesource(1);
         }
     }
     if(runtime->channels == 8){
@@ -217,6 +219,7 @@ static int aml_dai_i2s_trigger(struct snd_pcm_substream *substream, int cmd,
 				audio_out_i2s_enable(1);
 				if(amaudio2_enable == 1){
 					audio_hw_958_enable(1);
+					set_i2s_iec958_samesource(1);
 				}
 			}else{
 				audio_in_i2s_enable(1);
@@ -233,6 +236,7 @@ static int aml_dai_i2s_trigger(struct snd_pcm_substream *substream, int cmd,
 				audio_out_i2s_enable(0);
 				if(amaudio2_enable == 1){
 					audio_hw_958_enable(0);
+					set_i2s_iec958_samesource(0);
 				}
 			}else{
 				audio_in_i2s_enable(0);
-- 
2.19.0

